src/docfx/lib/path/PackageUrl.cs

```diff
@@ -0,0 +1,64 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using Newtonsoft.Json;
+
+namespace Microsoft.Docs.Build
+{
+    /// <summary>
+    /// Define the location of a package.
+    /// A package is any of the following:
+    ///   a) A folder
+    ///   b) A git url in the form of `{remote_url}#{commit-ish}`, when cloned, results in a)
+    /// The commit-sh can be any tag, sha, or branch. The default commit-ish is master.
+    /// </summary>
+    [JsonConverter(typeof(ShortHandConverter))]
+    internal readonly struct PackageUrl
```

Comment:
PackageUrl is too general, especially we also have files to download？

Reply:
since it's related to git, how about GitPackageUrl?

Reply:
A package is just a folder of files, git is one way to retrieve these files, we may in the future support other types of packages like zip files.

The idea came from https://docs.npmjs.com/about-packages-and-modules#about-package-formats

Reply:
But this PackageUrl has "Remote" and "Branch"? Is that common for all kinds of resource we want to restore?

Reply:
This is a union type, Remote and Branch only has value if the type is Git, Path only has value if the type is Folder, C# currently don't have a good way to express union types

Reply:
https://github.com/dotnet/csharplang/issues/113 for a detailed discussion around discriminated unions